## **TASK 3 - CLEANING DATA**

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df3 = pd.read_csv('/content/us-shein-mens_clothes-1891.csv')
df3.head()
df3.info()
df3.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1890 entries, 0 to 1889
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   color-count          1141 non-null   float64
 1   goods-title-link     1889 non-null   object 
 2   selling_proposition  1277 non-null   object 
 3   price                1889 non-null   object 
 4   discount             1860 non-null   object 
 5   rank-title           560 non-null    object 
 6   rank-sub             560 non-null    object 
dtypes: float64(1), object(6)
memory usage: 103.5+ KB


,0
color-count,749
goods-title-link,1
selling_proposition,613
price,1
discount,30
rank-title,1330
rank-sub,1330


In [2]:
quality_report = pd.DataFrame({
    'Nulls': df3.isnull().sum(),
    'Null %': (df3.isnull().sum() / len(df3) * 100).round(1),
    'Dtype': df3.dtypes,
    'Unique Values': df3.nunique()
})
print(quality_report)
print(f"\nDuplicate rows: {df3.duplicated().sum()}")
print(f"Total rows: {len(df3)}")

                     Nulls  Null %    Dtype  Unique Values
color-count            749    39.6  float64             19
goods-title-link         1     0.1   object           1631
selling_proposition    613    32.4   object             87
price                    1     0.1   object            604
discount                30     1.6   object             67
rank-title            1330    70.4   object             19
rank-sub              1330    70.4   object            376

Duplicate rows: 86
Total rows: 1890


In [3]:
print(df3['price'].head(10))
print(df3['discount'].head(10))

0    $11.99
1    $11.19
2     $8.09
3     $9.99
4     $3.48
5    $24.69
6    $22.79
7     $4.10
8     $7.99
9    $23.95
Name: price, dtype: object
0     -7%
1    -13%
2    -13%
3    -12%
4    -70%
5    -12%
6     -7%
7    -56%
8    -13%
9    -14%
Name: discount, dtype: object


In [4]:
print(df3['price'].head(10))
print(df3['discount'].head(10))

0    $11.99
1    $11.19
2     $8.09
3     $9.99
4     $3.48
5    $24.69
6    $22.79
7     $4.10
8     $7.99
9    $23.95
Name: price, dtype: object
0     -7%
1    -13%
2    -13%
3    -12%
4    -70%
5    -12%
6     -7%
7    -56%
8    -13%
9    -14%
Name: discount, dtype: object


In [5]:
# Clean price
df3['price'] = df3['price'].astype(str).str.replace('$', '', regex=False)
df3['price'] = pd.to_numeric(df3['price'], errors='coerce')

# Clean discount
df3['discount'] = df3['discount'].astype(str).str.replace('%', '', regex=False).str.replace('-', '', regex=False)
df3['discount'] = pd.to_numeric(df3['discount'], errors='coerce')

df3[['price', 'discount']].head(10)
df3[['price', 'discount']].describe()

,price,discount
count,1889.000000,1860.000000
mean,12.294087,18.246237
std,5.766049,14.584237
min,2.800000,5.000000
25%,8.390000,8.000000
50%,11.080000,13.000000
75%,15.090000,18.000000
max,61.290000,87.000000


In [6]:

df3['price'] = df3['price'].fillna(df3['price'].median())

df3['discount'] = df3['discount'].fillna(0)

df3['color-count'] = df3['color-count'].fillna(1)

df3['selling_proposition'] = df3['selling_proposition'].fillna('None')

df3['rank-title'] = df3['rank-title'].fillna('Not Ranked')

df3['rank-sub'] = df3['rank-sub'].fillna('Not Ranked')

df3 = df3.dropna(subset=['goods-title-link'])

df3.isnull().sum()

,0
color-count,0
goods-title-link,0
selling_proposition,0
price,0
discount,0
rank-title,0
rank-sub,0


**Missing Data Strategy:**
- Price (1 null): Filled with median — negligible impact, avoids losing a row.
- Discount (30 nulls): Filled with 0 — absence of a discount value likely means no promotion was applied.
- Color-count (749 nulls): Filled with 1 — likely represents single-variant products where color count wasn't tracked.
- Selling proposition (613 nulls): Filled with 'None' — this is an optional marketing tag; absence is meaningful, not an error.
- Rank-title/Rank-sub (~1330 nulls each): Filled with 'Not Ranked' — most products aren't part of a bestseller ranking list.
- Goods-title-link (1 null): Row dropped — this is the core product identifier; can't be meaningfully imputed.

In [7]:
dupes_before = df3.duplicated().sum()
df3 = df3.drop_duplicates()
print(f"Duplicates removed: {dupes_before}")
print(f"Remaining rows: {len(df3)}")

Duplicates removed: 86
Remaining rows: 1803


In [8]:
print(df3['selling_proposition'].unique())
print(df3['rank-title'].unique())

['10k+ sold recently' 'None' '30+ sold recently' '80+ sold recently'
 '3.8k+ sold recently' '100+ sold recently' '500+ sold recently'
 '600+ sold recently' '800+ sold recently' '10+ sold recently'
 '1.0k+ sold recently' '1.3k+ sold recently' '20+ sold recently'
 '6.8k+ sold recently' '1.2k+ sold recently' '1.6k+ sold recently'
 '3.0k+ sold recently' '6.3k+ sold recently' '300+ sold recently'
 '1.7k+ sold recently' '3.2k+ sold recently' '4.7k+ sold recently'
 '1.8k+ sold recently' '6.0k+ sold recently' '5.9k+ sold recently'
 '50+ sold recently' '2.3k+ sold recently' '40+ sold recently'
 '3.9k+ sold recently' '7.6k+ sold recently' '6.5k+ sold recently'
 '2.6k+ sold recently' '200+ sold recently' '5.3k+ sold recently'
 '1.1k+ sold recently' '4.3k+ sold recently' '5.0k+ sold recently'
 '1.5k+ sold recently' '900+ sold recently' '6.2k+ sold recently'
 '6.1k+ sold recently' '2.7k+ sold recently' '3.3k+ sold recently'
 '2.8k+ sold recently' '2.9k+ sold recently' '2.5k+ sold recently'
 '3.6k+ 

In [9]:
def detect_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower) | (df[column] > upper)]
    print(f"{column}: {len(outliers)} outliers (bounds: {lower:.2f} to {upper:.2f})")
    return outliers

price_outliers = detect_outliers_iqr(df3, 'price')
discount_outliers = detect_outliers_iqr(df3, 'discount')

price: 42 outliers (bounds: -1.67 to 25.15)
discount: 279 outliers (bounds: -9.50 to 34.50)


**Outlier Decision:** Price outliers (42 items, above $25.15) and discount
outliers (279 items, above 34.5%) were both retained rather than removed.
These represent genuine premium-priced products and legitimate deep-discount
promotions respectively — common and expected in retail/flash-sale data —
rather than data entry errors. Removing them would discard real business
signal.

In [10]:
df3.dtypes

,0
color-count,float64
goods-title-link,object
selling_proposition,object
price,float64
discount,float64
rank-title,object
rank-sub,object


In [11]:

print("Duplicates before removal:", df3.duplicated().sum())

df3 = df3.drop_duplicates()


print("Duplicates after removal:", df3.duplicated().sum())
print("Row count after removal:", len(df3))

Duplicates before removal: 0
Duplicates after removal: 0
Row count after removal: 1803


In [12]:
summary = pd.DataFrame({
    'Metric': ['Row Count', 'Null Count (total)', 'Duplicate Count', 'Correct Dtypes'],
    'Before Cleaning': [1889, 4123, 86, '1 (price/discount as object)'],
    'After Cleaning': [len(df3), df3.isnull().sum().sum(), df3.duplicated().sum(), 'All 7 columns correct']
})
summary

,Metric,Before Cleaning,After Cleaning
0,Row Count,1889,1803
1,Null Count (total),4123,0
2,Duplicate Count,86,0
3,Correct Dtypes,1 (price/discount as object),All 7 columns correct


In [13]:
df3.to_csv('/content/us-shein-mens_clothes-1891_cleaned.csv', index=False)

In [14]:
df3.to_csv('/content/us-shein-mens_clothes-1891_cleaned.csv', index=False)
print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


In [15]:
from google.colab import files
files.download('/content/us-shein-mens_clothes-1891_cleaned.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>